src/main/scala/org/apache/spark/rdd/OrderedRDD.scala

```diff
@@ -0,0 +1,154 @@
+package org.apache.spark.rdd
+
+import org.apache.spark._
+import org.broadinstitute.hail.Utils._
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+import scala.util.hashing._
+
+object OrderedRDD {
+  def apply[T, K, V](rdd: RDD[(K, V)], reducedRDD: Option[RDD[K]] = None, check: Boolean = true)
+    (implicit ev: (K) => T, tOrd: Ordering[T], kOrd: Ordering[K], tct: ClassTag[T], kct: ClassTag[K]): OrderedRDD[T, K, V] = {
+    rdd match {
+      case _: OrderedRDD[T, K, V] =>
+        rdd.asInstanceOf[OrderedRDD[T, K, V]]
+      case _: EmptyRDD[(K, V)] => new OrderedRDD(rdd, OrderedPartitioner(Array.empty[T]))
+      case _ =>
+        rdd.partitioner match {
+          case Some(p: OrderedPartitioner[T, K]) =>
+            new OrderedRDD(rdd, p)
+          case _ =>
+            val coercedRDD: Option[OrderedRDD[T, K, V]] = if (check) {
+              val res = reducedRDD.getOrElse(rdd.map(_._1)).mapPartitionsWithIndex { case (index, it) =>
+                if (it.isEmpty)
+                  Iterator((index, true, true, None))
+                else {
+                  var sortedT = true
+                  var sortedK = true
+                  var continue = true
+                  val firstK = it.next()
+                  var lastK = firstK
+                  var lastT = ev(firstK)
+
+                  while (it.hasNext && continue) {
+                    val k = it.next()
+                    val t = ev(k)
+                    if (tOrd.lt(t, lastT)) {
+                      sortedT = false
+                      continue = false
+                    } else if (kOrd.lt(k, lastK))
+                      sortedK = false
+
+                    lastK = k
+                    lastT = t
+                  }
```

Comment:
I think this case-lambda is sufficiently complex to be extracted into a helper function that perhaps has a test of its own.